### PR TITLE
Private recordings

### PR DESCRIPTION
**Pull request by jaril**
This is the first patch to add private recordings to Replay. The private aspect itself is simple for now - a private recording can only be viewed by the user who created the recording. These new private recording features are under a pref, so set `app.features.private = true` to see the changes.

This patch specifically:
- Infers whether a user can view a recording or not based on Hasura's response to a query for getting the recording
- Shows an error screen to a user if a recording is private and they're either 1) not logged in, or 2) not the user who created the recording.
- Allows assignment of public/private to a recording in `/view` through the dropdown, or clicking the Public/Private label

### Recordings page
image

### Error states
image
image

### CODE REVIEW

**Comment by jaril** [comment]
Next steps:
- Add support for collaborators
- Allow users to toggle public/private while looking at a recording
- UI polish for the toggle public/private label in the recordings page

File: src/ui/components/DevTools.js [hunk 1]
```diff
@@ -47,34 +60,53 @@ function getUploadingMessage(uploading) {
...
-        <Comments />
-        {hasFocusedComment && <div className="app-mask" onClick={unfocusComment} />}
-        <DevtoolsSplitBox tooltip={tooltip} updateTimelineDimensions={updateTimelineDimensions} />
-      </>
-    );
+  if (queryIsLoading) {
+    return <Loader />;
   }
+
+  const isAuthorized = features.private ? getIsAuthorized({ data, error, isAuthenticated }) : true;
```
Comment (jasonLaster) [inline src/ui/components/DevTools.js:91]: nit: lets do the `features.private` check in getIsAuthorized

File: src/ui/components/DevTools.js [hunk 2]
```diff
@@ -47,34 +60,53 @@ function getUploadingMessage(uploading) {
...
-      updateTimelineDimensions,
-      recordingDuration,
-    } = this.props;
+function getIsAuthorized({ data, error, isAuthenticated }) {
+  // We let Hasura decide whether or not the user can view a recording. The response to our query
+  // will have a recording if they're authorized to view the recording, and will be empty if not.
+  // What this doesn't explicitly tell us is *why* that user is allowed to view the recording: is the
+  // recording public, or is the user the original creator for that recording? You can check that
+  // by checking the recording's user field, which will be null for non-creator users.
+  return data.recordings.length === 0 ? false : true;
```
Comment (jasonLaster) [inline src/ui/components/DevTools.js:69]: just a personal thing you can return `data.recordings.length > 0`

File: src/ui/components/Recordings/Recording.js [hunk 3]
```diff
@@ -2,12 +2,35 @@ import React, { useState } from "react";
 import Title from "../shared/Title";
 import Dropdown from "devtools/client/debugger/src/components/shared/Dropdown";
 import moment from "moment";
+import { gql, useMutation } from "@apollo/client";
+import { features } from "ui/utils/prefs";
+
+const UPDATE_IS_PRIVATE = gql`
+  mutation MyMutation($recordingId: String, $isPrivate: Boolean) {
```
Comment (jasonLaster) [inline src/ui/components/Recordings/Recording.js:9]: lets name our mutations `SetRecordingIsPrivate`

File: src/ui/components/Recordings/Recording.js [hunk 4]
```diff
@@ -18,28 +41,46 @@ const DropdownPanel = ({ editingTitle, setEditingTitle, onDeleteRecording, recor
       <div className="menu-item" onClick={() => onDeleteRecording(recordingId)}>
         Delete Recording
       </div>
+      {!features.private ? null : isPrivate ? (
+        <div className="menu-item" onClick={toggleIsPrivate}>
+          Make recording public
```
Comment (jasonLaster) [inline src/ui/components/Recordings/Recording.js:46]: Generally we can assume context with copy `Make public` vs `Make recording public`

**Review by jasonLaster — APPROVED** [review]
Looking good. I couldn't test because I do not have recordings, but I feel like this work is safe to land and build on

**Comment by jaril** [comment]
[Loom Demo](https://www.loom.com/share/2ca1b1e0a8ae4a759cf5dc85ddc2c1da) with the `app.features.private = true`. Results are:

user | public recording | private recording
--- | --- | ---
logged-in author | ✅ shown |  ✅ shown 
logged-in non-author | ✅ shown | ❌ hidden
logged-out anonymous | ✅ shown | ❌ hidden

[Loom Demo](https://www.loom.com/share/f01d162c4ddc4a69b5339d0329572ef2) with `app.features.private = false`. Results are:

user | public recording | private recording
--- | --- | ---
logged-in author | ✅ shown |  ✅ shown 
logged-in non-author | ✅ shown | ✅ shown 
logged-out anonymous | ✅ shown | ✅ shown

**Pull request merged**